File: WebHuntiOS/WebHunt/ViewModels/ProfileViewModel.swift
import Foundation

@MainActor
final class ProfileViewModel: ObservableObject {
    @Published var profileUser: User?
    @Published var isLoading = false
    @Published var errorMessage: String?

    private let api = APIClient.shared

    var userProducts: [Product] {
        guard let user = profileUser, let products = user.products else { return [] }
        return Array(products.values).sorted { $0.id > $1.id }
    }

    var upvotedProducts: [Product] {
        guard let user = profileUser, let products = user.upvotedProducts else { return [] }
        return Array(products.values).sorted { $0.id > $1.id }
    }

    func fetchProfile(username: String) async {
        isLoading = true
        errorMessage = nil

        do {
            profileUser = try await api.fetchUserByUsername(username: username)
        } catch let error as APIError {
            errorMessage = error.errorDescription
        } catch {
            errorMessage = error.localizedDescription
        }

        isLoading = false
    }

    func fetchProfile(userId: Int) async {
        isLoading = true
        errorMessage = nil

        do {
            profileUser = try await api.fetchUser(id: userId)
        } catch let error as APIError {
            errorMessage = error.errorDescription
        } catch {
            errorMessage = error.localizedDescription
        }

        isLoading = false
    }
}

File: WebHuntAndroid/app/src/main/java/dev/webhunt/android/ui/profile/ProfileViewModel.kt
package dev.webhunt.android.ui.profile

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import dev.webhunt.android.data.model.Product
import dev.webhunt.android.data.model.User
import dev.webhunt.android.data.network.ApiClient
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class ProfileViewModel @Inject constructor(
    private val apiClient: ApiClient
) : ViewModel() {

    var profileUser by mutableStateOf<User?>(null)
        private set
    var isLoading by mutableStateOf(false)
        private set
    var errorMessage by mutableStateOf<String?>(null)
        private set

    val userProducts: List<Product>
        get() {
            val products = profileUser?.products ?: return emptyList()
            return products.values.sortedByDescending { it.id }
        }

    val upvotedProducts: List<Product>
        get() {
            val products = profileUser?.upvotedProducts ?: return emptyList()
            return products.values.sortedByDescending { it.id }
        }

    fun fetchProfile(username: String) {
        viewModelScope.launch {
            isLoading = true
            errorMessage = null
            try {
                profileUser = apiClient.fetchUserByUsername(username)
            } catch (e: Exception) {
                errorMessage = e.message ?: "Failed to load profile"
            }
            isLoading = false
        }
    }

    fun fetchProfile(userId: Int) {
        viewModelScope.launch {
            isLoading = true
            errorMessage = null
            try {
                profileUser = apiClient.fetchUser(userId)
            } catch (e: Exception) {
                errorMessage = e.message ?: "Failed to load profile"
            }
            isLoading = false
        }
    }

    fun resolveUrl(path: String?): String? = apiClient.resolveUrl(path)
}
